lib/algo/manifold_interp/PCHIPInterpolator.cpp

```diff
@@ -0,0 +1,219 @@
+/******************************************************************************
+ *
+ * Copyright (c) 2013-2024, Lawrence Livermore National Security, LLC
+ * and other libROM project developers. See the top-level COPYRIGHT
+ * file for details.
+ *
+ * SPDX-License-Identifier: (Apache-2.0 OR MIT)
+ *
+ *****************************************************************************/
+
+/**
+ * Implements PCHIP algorithm.  Based on "A METHOD FOR CONSTRUCTING LOCAL MONOTONE
+ * PIECEWISE CUBIC INTERPOLANTS", Fritsch and Butland (1984).  as well as "MONOTONE
+ * PIECEWISE CUBIC INTERPOLATION," Fritsch and Carlson (1980)
+ *
+ */
+#include "PCHIPInterpolator.h"
+
+#include <cfloat>
+#include <limits.h>
+#include <cmath>
+#include "linalg/Vector.h"
+#include "mpi.h"
+
+/* Use C++11 built-in shared pointers if available; else fallback to Boost. */
+#if __cplusplus >= 201103L
+#include <memory>
+#else
+#include <boost/shared_ptr.hpp>
+#endif
+using namespace std;
+
+namespace CAROM {
+
+void PCHIPInterpolator::interpolate(std::vector<Vector*>& snapshot_ts,
+                                    std::vector<Vector*>& snapshots,
+                                    std::vector<Vector*>& output_ts,
+                                    std::vector<Vector*>& output_snapshots)
+{
+    CAROM_VERIFY(snapshot_ts.size() == snapshots.size());
+    CAROM_VERIFY(snapshot_ts.size() > 0);
+    CAROM_VERIFY(output_ts.size() > 0);
+    CAROM_VERIFY(output_snapshots.size() == 0);
+    CAROM_VERIFY(snapshot_ts[0]->getData()[0] - FLT_EPSILON <=
+                 output_ts[0]->getData()[0]  &&
+                 output_ts[output_ts.size()-1]->getData()[0] <= snapshot_ts[snapshot_ts.size()
+                         -1]->getData()[0] + FLT_EPSILON);
+
+    int n_out = output_ts.size();
+    int n_snap = snapshots.size();
+    int n_dim = snapshots[0]->dim();
+
+    for(int i = 0; i < n_out; ++i)
+    {
+        Vector* temp_snapshot = new Vector(snapshots[0]->dim(),
+                                           snapshots[0]->distributed());
+        output_snapshots.push_back(temp_snapshot);
+    }
+
+    for(int i = 0; i < n_dim; ++i)
+    {
+        double h_temp,delta_temp, t;
+        std::vector<double> h,d,delta,t_in,y_in;
+        for(int j = 0; j < n_snap-1; ++j)
+        {
+            h_temp = snapshot_ts[j+1]->getData()[0] - snapshot_ts[j]->getData()[0];
+            h.push_back(h_temp);
+            delta_temp = (snapshots[j+1]->getData()[i] - snapshots[j]->getData()[i])/h_temp;
+            delta.push_back(delta_temp);
+            t_in.push_back(snapshot_ts[j]->getData()[0]);
+            y_in.push_back(snapshots[j]->getData()[i]);
+        }
+        t_in.push_back(snapshot_ts[n_snap-1]->getData()[0]);
+        y_in.push_back(snapshots[n_snap-1]->getData()[i]);
+
+        double d_temp = ((2*h[0] + h[1])*delta[0] - h[0]*delta[1])/(h[0]+h[1]);
```

Comment:
This is probably minor but just to note that if some input times are equal and not strictly increasing, `/ (h[0] + h[1])` might divide by 0?

Reply:
Added a verify statement to guarantee that snapshot_ts are strictly increasing.